src/main/java/frc/robot/subsystems/swerve/module/Module.java

```diff
@@ -0,0 +1,310 @@
+import frc.robot.subsystems.swerve.module.records.SteerSignals;
+import frc.utils.Conversions;
+import frc.utils.calibration.sysid.SysIdCalibrator;
+import org.littletonrobotics.junction.Logger;
+
+import java.util.Arrays;
+
+public class Module {
+
+	private final ModuleSpecificConstants constants;
+
+	private final IAngleEncoder encoder;
+	private final EncoderSignals encoderSignals;
+
+	private final ControllableMotor steer;
+	private final SteerRequests steerRequests;
+	private final SteerSignals steerSignals;
+
+	private final ControllableMotor drive;
+	private final DriveRequests driveRequests;
+	private final DriveSignals driveSignals;
+
+	private final ModuleInputsAutoLogged moduleInputs;
+	private final DriveInputsAutoLogged driveInputs;
+	private final DriveCouplingInputsAutoLogged driveCouplingInputs;
+
+	private SwerveModuleState targetState;
+	private Rotation2d startingSteerPosition;
+	private boolean isClosedLoop;
+
+	public Module(
+		ModuleSpecificConstants constants,
+		IAngleEncoder encoder,
+		EncoderSignals encoderSignals,
+		ControllableMotor steer,
+		SteerRequests steerRequests,
+		SteerSignals steerSignals,
+		ControllableMotor drive,
+		DriveRequests driveRequests,
+		DriveSignals driveSignals
+	) {
+		this.constants = constants;
+
+		this.encoder = encoder;
+		this.encoderSignals = encoderSignals;
+
+		this.steer = steer;
+		this.steerRequests = steerRequests;
+		this.steerSignals = steerSignals;
+
+		this.drive = drive;
+		this.driveRequests = driveRequests;
+		this.driveSignals = driveSignals;
+
+		this.targetState = new SwerveModuleState();
+		this.startingSteerPosition = new Rotation2d();
+		this.isClosedLoop = ModuleConstants.DEFAULT_IS_CLOSE_LOOP;
+		this.moduleInputs = new ModuleInputsAutoLogged();
+		this.driveInputs = new DriveInputsAutoLogged();
+		this.driveCouplingInputs = new DriveCouplingInputsAutoLogged();
+
+		updateInputs();
+		resetSteerByEncoder();
+	}
+
+	public SysIdCalibrator.SysIdConfigInfo getSteerSysIdConfigInfo() {
+		return steer.getSysidConfigInfo();
+	}
+
+	public SysIdCalibrator.SysIdConfigInfo getDriveSysIdConfigInfo() {
+		return drive.getSysidConfigInfo();
+	}
+
+	public Translation2d getPositionFromCenterMeters() {
+		return constants.positionFromCenterMeters();
+	}
+
+	public double toDriveMeters(Rotation2d angle) {
+		return Conversions.angleToDistance(angle, constants.wheelDiameterMeters());
+	}
+
+
+	private void fixDriveInputsCoupling() {
+		driveCouplingInputs.uncoupledVelocityAnglesPerSecond = ModuleUtils
+			.uncoupleDriveAngle(driveSignals.velocity().getLatestValue(), steerSignals.velocity().getLatestValue(), constants.couplingRatio());
+
+		driveCouplingInputs.uncoupledPositions = new Rotation2d[driveSignals.position().asArray().length];
+		for (int i = 0; i < driveCouplingInputs.uncoupledPositions.length; i++) {
+			Rotation2d steerDelta = Rotation2d
+				.fromRotations(steerSignals.position().asArray()[i].getRotations() - startingSteerPosition.getRotations());
```

Comment:
make this a util function

Reply:
all rotation2d util will be in another PR